### PR TITLE
Refactor GraphQL Server and CreateYoga to Support "api serve" with Fastify Server

### PR DESCRIPTION
**Pull request by dthyresson**
As part of the [effort to have fastify plugins for running the redwood api and web servers](https://github.com/redwoodjs/redwood/pull/8119), this PR refactors parts of the graphql-server package so it too can be used in a more "serverful" manner.

And, in doing so, begin the support for "RedwoodJS Realtime" via GraphQL subscriptions and Live Queries when deployed with a stateful server and memory stores.

This PR:

* creates of the `createYoga` portion on the createGraphQLHandler so it can be used by both the serverless function `graphql` and a 
* new Fastify "plugin" that creates a server and registers yoga as a handler
* adds a new `allowGraphiQL` option because when running even in dev as `yarn rw api serve` this is considered production and both introspection and the playground are disabled. Now, you can control both and allow the playground to be used with `rw ap serve`
* updates the server file template to include a yoga server to experiment with 

## Use

The `server.ts` here creates a server and also sets up subscription and live query support (in the playground only, there is no web side support yet to actually listen or subscribe).

Currently, the types and resolvers are added to the RedwoodJS GraphQL schema via the extra `schemaOptions` option when merging the schema from the sdls, services and directives.

```ts
// api/src/server.ts

import path from 'path'

import { useLiveQuery } from '@envelop/live-query'
import chalk from 'chalk'
import { config } from 'dotenv-defaults'
import Fastify from 'fastify'
import { OperationTypeNode } from 'graphql'

import {
  coerceRootPath,
  redwoodFastifyWeb,
  redwoodFastifyAPI,
  redwoodFastifyGraphQLServer,
  DEFAULT_REDWOOD_FASTIFY_CONFIG,
} from '@redwoodjs/fastify'
import { getPaths, getConfig } from '@redwoodjs/project-config'

import directives from 'src/directives/**/*.{js,ts}'
import sdls from 'src/graphql/**/*.sdl.{js,ts}'
import services from 'src/services/**/*.{js,ts}'

import { liveQueryStore, liveSchema } from './graphql/liveSchema'
import { subscriptions, pubSub } from './graphql/subscriptions'
import { logger } from './lib/logger'

async function serve() {
  // Load .env files
  const redwoodProjectPaths = getPaths()
  const redwoodConfig = getConfig()

  const apiRootPath = coerceRootPath(redwoodConfig.web.apiUrl)
  const port = redwoodConfig.web.port

  const tsServer = Date.now()

  config({
    path: path.join(redwoodProjectPaths.base, '.env'),
    defaults: path.join(redwoodProjectPaths.base, '.env.defaults'),
    multiline: true,
  })

  console.log(chalk.italic.dim('Starting API and Web Servers...'))

  // Configure Fastify
  const fastify = Fastify({
    ...DEFAULT_REDWOOD_FASTIFY_CONFIG,
  })

  await fastify.register(redwoodFastifyWeb)

  await fastify.register(redwoodFastifyAPI, {
    redwood: {
      apiRootPath,
    },
  })

  await fastify.register(redwoodFastifyGraphQLServer, {
    loggerConfig: {
      logger: logger,
      options: { query: true, data: true, level: 'trace' },
    },
    graphiQLEndpoint: '/yoga',
    sdls,
    services,
    directives,
    allowedOperations: [
      OperationTypeNode.SUBSCRIPTION,
      OperationTypeNode.QUERY,
      OperationTypeNode.MUTATION,
    ],
    allowIntrospection: true,
    allowGraphiQL: true,
    schemaOptions: { ...subscriptions, ...liveSchema },
    context: { pubSub },
    extraPlugins: [useLiveQuery({ liveQueryStore })],
  })

  // Start

  fastify.listen({ port })

  fastify.ready(() => {
    console.log(chalk.italic.dim('Took ' + (Date.now() - tsServer) + ' ms'))
    const on = chalk.magenta(`http://localhost:${port}${apiRootPath}`)
    const webServer = chalk.green(`http://localhost:${port}`)
    const apiServer = chalk.magenta(`http://localhost:${port}`)
    console.log(`Web server started on ${webServer}`)
    console.log(`API serving from ${apiServer}`)
    console.log(`API listening on ${on}`)
    const graphqlEnd = chalk.magenta(`${apiRootPath}graphql`)
    console.log(`GraphQL serverless function endpoint at ${graphqlEnd}`)
  })

  process.on('exit', () => {
    fastify.close()
  })
}

serve()
```

Note: I just realized that this example doesn't mix both LQ and Subs due to the way I set the schemaOptions. But both do work independently. Hence, why need a way to merge these into there schema in the next iteration.

Important here is to "allow" subscriptions, introspection and graphiql.

```ts
// api/src/graphql/liveSchema.ts

import { astFromDirective } from '@graphql-tools/utils'
import { GraphQLLiveDirective } from '@n1ru4l/graphql-live-query'
import { InMemoryLiveQueryStore } from '@n1ru4l/in-memory-live-query-store'

export const liveQueryStore = new InMemoryLiveQueryStore()

type Bid = {
  amount: number
}

type Auction = {
  id: string
  title: string
  bids: Bid[]
}

const auctions: Auction[] = [
  { id: '1', title: 'Digital-only PS5', bids: [{ amount: 100 }] },
]

export const liveSchema = {
  typeDefs: [
    /* GraphQL */ `
      type Query {
        auction(id: ID!): Auction
      }
      type Auction {
        id: ID!
        title: String!
        highestBid: Bid
        bids: [Bid!]!
      }
      type Bid {
        amount: Int!
      }
      type Mutation {
        bid(input: BidInput!): Bid
      }
      input BidInput {
        auctionId: ID!
        amount: Int!
      }
    `,
    astFromDirective(GraphQLLiveDirective),
  ],
  resolvers: {
    Query: {
      auction: (_, { id }) => auctions.find((a) => a.id === id),
    },
    Mutation: {
      bid: async (_, { input }) => {
        const { auctionId, amount } = input

        const index = auctions.findIndex((a) => a.id === auctionId)

        const bid = { amount }

        auctions[index].bids.push(bid)

        liveQueryStore.invalidate(`Auction:${auctionId}`)

        return bid
      },
    },
    Auction: {
      highestBid: ({ bids }: Auction) => {
        const [max] = bids.sort((a, b) => b.amount - a.amount)

        return max
      },
    },
  },
}
```

and Subscriptions

```ts
// api/src/graphql/subscriptions.ts

import { createPubSub } from '@graphql-yoga/node'

export const pubSub = createPubSub<{
  newMessage: [payload: { from: string; body: string }]
}>()

export const subscriptions = {
  typeDefs: [
    /* GraphQL */ `
      type Subscription {
        countdown(from: Int!, interval: Int!): Int!
      }
      type Query {
        room(id: ID!): [Message!]!
      }

      type Mutation {
        send(input: SendMessageInput!): Message!
      }

      type Subscription {
        newMessage(roomId: ID!): Message!
      }

      type Message {
        from: String
        body: String
      }

      input SendMessageInput {
        roomId: ID!
        from: String!
        body: String!
      }
    `,
  ],
  resolvers: {
    Query: {
      room: () => [],
    },
    Mutation: {
      send: (_, { input }, { pubSub }) => {
        const { roomId, ...newMessage } = input
        console.log('newMessage', newMessage)
        console.log('roomId', roomId)

        pubSub.publish('newMessage', roomId, newMessage)

        return newMessage
      },
    },
    Subscription: {
      countdown: {
        async *subscribe(_, { from, interval }) {
          for (let i = from; i >= 0; i--) {
            await new Promise((resolve) =>
              setTimeout(resolve, interval ?? 1000)
            )
            yield { countdown: i }
          }
        },
      },
      newMessage: {
        subscribe: (_, { roomId }, { pubSub }) =>
          pubSub.subscribe('newMessage', roomId),
        resolve: (payload) => payload,
      },
    },
  },
}
```

## Next Steps

There are many next steps to support realtime and GraphQL Subscriptions and Live Queries.

* figure out web side needs for Apollo Client
* What is the DX and how best to organize Subscription and LiveQuery types and resolvers 
*

### CODE REVIEW

**Comment by dthyresson** [comment]
Have started the work to merge in project subscriptions into the over schema and resolvers.

**Comment by dthyresson** [comment]
I have subscriptions and live queries running in a test app with this PR.

## Server File Setup

```ts
import path from 'path'

import { useLiveQuery } from '@envelop/live-query'
import { astFromDirective } from '@graphql-tools/utils'
import { GraphQLLiveDirective } from '@n1ru4l/graphql-live-query'
import chalk from 'chalk'
import { config } from 'dotenv-defaults'
import Fastify from 'fastify'
import { OperationTypeNode } from 'graphql'

import {
  coerceRootPath,
  redwoodFastifyWeb,
  redwoodFastifyAPI,
  redwoodFastifyGraphQLServer,
  DEFAULT_REDWOOD_FASTIFY_CONFIG,
} from '@redwoodjs/fastify'
import { getPaths, getConfig } from '@redwoodjs/project-config'

import directives from 'src/directives/**/*.{js,ts}'
import sdls from 'src/graphql/**/*.sdl.{js,ts}'
import services from 'src/services/**/*.{js,ts}'
import subscriptions from 'src/subscriptions/**/*.{js,ts}'

import { liveQueryStore } from 'src/services/auctions/auctions'

import { logger } from './lib/logger'
async function serve() {
  // Load .env files
  const redwoodProjectPaths = getPaths()
  const redwoodConfig = getConfig()

  const apiRootPath = coerceRootPath(redwoodConfig.web.apiUrl)
  const port = redwoodConfig.web.port

  const tsServer = Date.now()

  config({
    path: path.join(redwoodProjectPaths.base, '.env'),
    defaults: path.join(redwoodProjectPaths.base, '.env.defaults'),
    multiline: true,
  })

  console.log(chalk.italic.dim('Starting API and Web Servers...'))

  // Configure Fastify
  const fastify = Fastify({
    ...DEFAULT_REDWOOD_FASTIFY_CONFIG,
  })

  await fastify.register(redwoodFastifyWeb)

  await fastify.register(redwoodFastifyAPI, {
    redwood: {
      apiRootPath,
    },
  })

  await fastify.register(redwoodFastifyGraphQLServer, {
    loggerConfig: {
      logger: logger,
      options: { query: true, data: true, level: 'trace' },
    },
    graphiQLEndpoint: '/yoga',
    sdls,
    services,
    directives,
    subscriptions,
    allowedOperations: [
      OperationTypeNode.SUBSCRIPTION,
      OperationTypeNode.QUERY,
      OperationTypeNode.MUTATION,
    ],
    allowIntrospection: true,
    allowGraphiQL: true,
    schemaOptions: { typeDefs: [astFromDirective(GraphQLLiveDirective)] },
    extraPlugins: [useLiveQuery({ liveQueryStore })],
  })

  // Start

  fastify.listen({ port })

  fastify.ready(() => {
    console.log(chalk.italic.dim('Took ' + (Date.now() - tsServer) + ' ms'))
    const on = chalk.magenta(`http://localhost:${port}${apiRootPath}`)
    const webServer = chalk.green(`http://localhost:${port}`)
    const apiServer = chalk.magenta(`http://localhost:${port}`)
    console.log(`Web server started on ${webServer}`)
    console.log(`API serving from ${apiServer}`)
    console.log(`API listening on ${on}`)
    const graphqlEnd = chalk.magenta(`${apiRootPath}graphql`)
    console.log(`GraphQL serverless function endpoint at ${graphqlEnd}`)
  })

  process.on('exit', () => {
    fastify.close()
  })
}

serve()
```

## Subscriptions

* put in `api/src/subscriptions` like you would Redwood Directives
* we have the concept of a redwoodSubscription and can make a generator for it like directives -- or g realtime and have the option to setup a subscription of live query perhaps 

A subscription file can be 

```ts
import gql from 'graphql-tag'

export const schema = gql`
  type Subscription {
    countdown(from: Int!, interval: Int!): Int!
  }
`

const countdown = {
  countdown: {
    async *subscribe(_, { from, interval }) {
      for (let i = from; i >= 0; i--) {
        await new Promise((resolve) => setTimeout(resolve, interval ?? 1000))
        yield { countdown: i }
      }
    },
  },
}

export default countdown
``` 

or for a `newMessage`

```ts
import { createPubSub } from '@graphql-yoga/node'
import gql from 'graphql-tag'

import { logger } from 'src/lib/logger'

export const pubSubNewMessage = createPubSub<{
  newMessage: [payload: { from: string; body: string }]
}>()

export const schema = gql`
  type Subscription {
    newMessage(roomId: ID!): Message!
  }
`

const newMessage = {
  newMessage: {
    subscribe: (_, { roomId }) => {
      logger.debug({ roomId }, 'newMessage subscription')

      return pubSubNewMessage.subscribe(roomId)
    },
    resolve: (payload) => {
      logger.debug({ payload }, 'newMessage subscription resolve')

      return payload
    },
  },
}

export default newMessage
```

## Define SDL types

Here the `rooms.sdl.ts` defines the Message type used in the subscription and the mutation to `send` a new message

```ts
export const schema = gql`
  type Message {
    from: String
    body: String
  }

  type Query {
    room(id: ID!): [Message!]! @skipAuth
  }

  input SendMessageInput {
    roomId: ID!
    from: String!
    body: String!
  }

  type Mutation {
    send(input: SendMessageInput!): Message! @skipAuth
  }
`
```


### Services

The `api/src/services/rooms/rooms` gets in in memory pubSub from the subscription and when sending a new message mutation, publishes onto pub sub

```ts
import { logger } from 'src/lib/logger'
import { pubSubNewMessage } from 'src/subscriptions/newMessage/newMessage'

export const room = ({ id }) => [id]

// important to be async!
export const send = async ({ input }) => {
  logger.debug({ input }, 'send input')

  const { roomId, ...newMessage } = input

  pubSubNewMessage.publish(roomId, newMessage)

  return input
}
```

File: packages/graphql-server/src/functions/graphql.ts [hunk 1]
```diff
@@ -58,155 +35,11 @@ export const createGraphQLHandler = ({
   armorConfig,
   allowedOperations,
   allowIntrospection,
+  allowGraphiQL,
   defaultError = 'Something went wrong.',
   graphiQLEndpoint = '/graphql',
```
Comment (dthyresson) [inline packages/graphql-server/src/functions/graphql.ts:40]: Note that I may add another option so can have graphql endpoint and playground graphiql at two different urls.

File: packages/graphql-server/src/functions/graphql.ts [hunk 2]
```diff
@@ -216,6 +49,28 @@ export const createGraphQLHandler = ({
 
     let lambdaResponse: APIGatewayProxyResult
 
+    const { yoga, logger } = createGraphQLYoga({
+      healthCheckId,
+      loggerConfig,
+      context,
+      getCurrentUser,
+      onException,
+      generateGraphiQLHeader,
```
Comment (dthyresson) [inline packages/graphql-server/src/functions/graphql.ts:58]: Note once studio is upon, removing this header

File: packages/cli/src/commands/experimental/templates/server.ts.template [hunk 3]
```diff
@@ -55,7 +82,7 @@ async function serve() {
     console.log(`API serving from ${apiServer}`)
     console.log(`API listening on ${on}`)
     const graphqlEnd = chalk.magenta(`${apiRootPath}graphql`)
-    console.log(`GraphQL endpoint at ${graphqlEnd}`)
+    console.log(`GraphQL serverless function endpoint at ${graphqlEnd}`)
```
Comment (jtoar) [inline packages/cli/src/commands/experimental/templates/server.ts.template:85]:
```suggestion
    console.log(`GraphQL function endpoint at ${graphqlEnd}`)
```

File: docs/docs/graphql.md [hunk 4]
```diff
@@ -1464,6 +1464,7 @@ export const handler = createGraphQLHandler({
   sdls,
   services,
   allowIntrospection: true, // 👈 enable introspection in all environments
+  allowGraphiQL: true, // 👈 enable introspection in all environments
```
Comment (dthyresson) [inline docs/docs/graphql.md:1467]:
```suggestion
  allowGraphiQL: true, // 👈 enable GraphiQL Playground in all environments
```

**Review by jtoar — APPROVED** [review]

**Comment by jtoar** [comment]
@dthyresson and I talked about this one; we're synced up on concerns and next steps and want to get this into main so development can keep going on other milestones.

**Comment by dthyresson** [comment]
woohoo

**Pull request merged**